File: go/proxy_server.go
package main

import (
	"fmt"
	"io"
	"net"
	"strings"
)

func handleClient(client net.Conn) {
	defer client.Close()

	// Read the request from the client
	request := make([]byte, 1024)
	client.Read(request)

	// Extract the host and port from the request
	lines := strings.Split(string(request), "\n")
	hostPort := strings.Split(lines[0], " ")[1]
	hostPort = strings.TrimPrefix(hostPort, "http://")
	hostPort = strings.Split(hostPort, "/")[0]

	// Split into host and port
	host := hostPort
	port := "80"
	if strings.Contains(hostPort, ":") {
		hostPortSplit := strings.Split(hostPort, ":")
		host = hostPortSplit[0]
		port = hostPortSplit[1]
	}

	// Connect to the remote server
	server, err := net.Dial("tcp", net.JoinHostPort(host, port))
	if err != nil {
		fmt.Println("Error connecting to server:", err)
		return
	}
	defer server.Close()

	// Send the request to the server
	server.Write(request)

	// Copy the server's response back to the client
	io.Copy(client, server)
}

func main() {
	// Listen on port 8888
	listener, err := net.Listen("tcp", ":8888")
	if err != nil {
		fmt.Println("Error starting proxy server:", err)
		return
	}
	defer listener.Close()

	fmt.Println("Proxy server listening on port 8888")

	for {
		client, err := listener.Accept()
		if err != nil {
			fmt.Println("Error accepting connection:", err)
			continue
		}

		go handleClient(client)
	}
}

File: java/proxy_server.java
import java.io.*;
import java.net.*;

public class ProxyServer {

    public static void main(String[] args) throws IOException {
        ServerSocket serverSocket = new ServerSocket(8888);
        System.out.println("Proxy server listening on port 8888");

        while (true) {
            Socket clientSocket = serverSocket.accept();
            new Thread(() -> handleClient(clientSocket)).start();
        }
    }

    public static void handleClient(Socket clientSocket) {
        try {
            BufferedReader inFromClient = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
            DataOutputStream outToClient = new DataOutputStream(clientSocket.getOutputStream());

            // Read the request from the client
            String requestLine = inFromClient.readLine();
            String[] tokens = requestLine.split(" ");
            String url = tokens[1];
            URL urlObj = new URL(url);
            String host = urlObj.getHost();
            int port = urlObj.getPort() == -1 ? 80 : urlObj.getPort();

            // Connect to the remote server
            Socket serverSocket = new Socket(host, port);
            BufferedReader inFromServer = new BufferedReader(new InputStreamReader(serverSocket.getInputStream()));
            DataOutputStream outToServer = new DataOutputStream(serverSocket.getOutputStream());

            // Send the request to the server
            outToServer.writeBytes(requestLine + "\r\n");
            String headerLine;
            while (!(headerLine = inFromClient.readLine()).equals("")) {
                outToServer.writeBytes(headerLine + "\r\n");
            }
            outToServer.writeBytes("\r\n");

            // Read the server's response and send it to the client
            String responseLine;
            while ((responseLine = inFromServer.readLine()) != null) {
                outToClient.writeBytes(responseLine + "\r\n");
            }

            serverSocket.close();
            clientSocket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
